packages/steps/src/BuildWorkflowValidator.ts

```diff
@@ -0,0 +1,81 @@
+import { BuildStep } from './BuildStep.js';
+import { BuildWorkflow } from './BuildWorkflow.js';
+import { BuildConfigError } from './errors/BuildConfigError.js';
+import { BuildWorkflowError } from './errors/BuildWorkflowError.js';
+import { uniq } from './utils/expodash/uniq.js';
+import { findOutputPaths } from './utils/template.js';
+
+export class BuildWorkflowValidator {
+  constructor(private readonly workflow: BuildWorkflow) {}
+
+  public validate(): void {
+    const errors: BuildConfigError[] = [];
+    errors.push(...this.validateUniqueStepIds());
+    errors.push(...this.validateInputs());
+    if (errors.length !== 0) {
+      throw new BuildWorkflowError('Build workflow is invalid', errors);
+    }
+  }
+
+  private validateUniqueStepIds(): BuildConfigError[] {
+    const stepIds = this.workflow.buildSteps.map(({ id }) => id);
+    const uniqueStepIds = uniq(stepIds);
+    if (stepIds.length === uniqueStepIds.length) {
+      return [];
+    }
+    const notUsedUniqueStepIdsSet = new Set(uniqueStepIds);
+    const duplicatedStepIdsSet = new Set<string>();
+    for (const stepId of stepIds) {
+      if (notUsedUniqueStepIdsSet.has(stepId)) {
+        notUsedUniqueStepIdsSet.delete(stepId);
+      } else {
+        duplicatedStepIdsSet.add(stepId);
+      }
+    }
```

Comment:
```suggestion
    const visitedStepIdsSet = new Set<string>();
    const duplicatedStepIdsSet = new Set<string>();
    for (const stepId of stepIds) {
      if (visitedStepIdsSet.has(stepId)) {
        duplicatedStepIdsSet.add(stepId);
      } else {
        visitedStepIdsSet.add(stepId);
      }
    }
```

No early exit means it's slower, but I think this simplifies the code.
If you prefer to keep your approach, consider renaming notUsedUniqueStepIdsSet